include/eosio/vm/x86_64.hpp

```diff
@@ -0,0 +1,2092 @@
+#pragma once
+
+#include <eosio/vm/allocator.hpp>
+#include <eosio/vm/types.hpp>
+
+#include <eosio/vm/softfloat.hpp>
+
+namespace eosio { namespace vm {
+
+
+
+   // Random notes:
+   // - branch instructions return the address that will need to be updated
+   // - label instructions return the address of the target
+   // - fix_branch will be called when the branch target is resolved
+   // - It would make everything more efficient to make RAX always represent the top of
+   //   the stack.
+   //
+   // - The base of memory is stored in rsi
+   //
+   // - FIXME: Factor the machine instructions into a separate assembler class.
+   template<typename Context>
+   class machine_code_writer {
+    public:
+      machine_code_writer(growable_allocator& alloc, std::size_t source_bytes, module& mod) :
```

Comment:
There's a couple of "magic" numbers in this function (40, 17, ...). Refactoring them into named constants might help with easier comprehension for anyone that has to do future maintenance.

Reply:
I'm not sure how to make it better.  I definitely don't want to move them out-of-line, because they represent the size in bytes of the code that immediately follows them.  They're already factored out as named constants.

Reply:
Ah yeah, I definitely see what you're saying. My confusion was just specifically around where those numbers come from, but that could likely just be due to not having the right background knowledge.

Reply:
This code is pretty opaque.  I just couldn't think of a better way to handle it.  I'll at least add some comments.